[opentitanlib, testutils] Add UART bitbanging encoding/decoding utilities

AlexJones0 opened the pull request:
Context: This is the first of a series of PRs to integrate bitbanging utilities with opentitanlib, such that existing transport peripherals (UART, SPI) can be replaced with bitbanged equivalents, with a goal of recording waves that can be replayed during provisioning.

This PR adds utilities for encoding and decoding bitbanged UART transmissions in the same style as the existing SPI/I2C/PWM bitbanging utilities. These are designed to be self-contained such that they could easily be used as a stand-alone crate outside of opentitanlib if desired.

Because UART receiving is asynchronous, the decoder is designed to be stateful such that it can any arbitrary partial number of samples of a UART transmission. This allows a receiver to begin (and later finish) decoding a UART transmission from partial samples, meaning that any implementation don't need to worry about potentially sampling mid-transmission.

The last commit features a variety of unit tests to check that the bitbanging utilities are working correctly.

engdoreis commented on sw/host/opentitanlib/src/test_utils/bitbanging/uart.rs:
```diff
@@ -118,18 +127,212 @@ impl<const TX: u8> UartBitbangEncoder<TX> {
+            DecodingState::Data { mut bits } => {
+                self.parsed_data |= rx << bits;
+                bits += 1;
+                if bits >= self.config.data_bits {
+                    if self.config.parity == Parity::None {
+                        self.state = DecodingState::Stop {
+                            data: 0x00,
+                            bits: 0,
+                        };
+                    } else {
+                        self.state = DecodingState::Parity;
+                    }
+                } else {
+                    self.state = DecodingState::Data { bits };
+                }
```
Nit.
```suggestion
         self.state = if bits >= self.config.data_bits {
                    if self.config.parity == Parity::None {
                        DecodingState::Stop {
                            data: 0x00,
                            bits: 0,
                        }
                    } else {
                        DecodingState::Parity
                    }
                } else {
                    DecodingState::Data { bits }
                }
```

AlexJones0 replied:
Good catch!

pamaury commented on sw/host/opentitanlib/src/test_utils/bitbanging/uart.rs:
```diff
@@ -0,0 +1,135 @@
+/// UART frame configuration to use for UART bitbanging. Assumes LSB first.
+#[derive(Clone, Debug)]
+pub struct UartBitbangConfig {
+    pub data_bits: u8, // Assumes <= 8 data bits
```
For `stop_bits`, I wonder if it should be an enum instead? With a `u8` you cannot even represent 1.5 so if that's needed in the future, we would have to change the type (although I don't know how common 1.5 bit stop is). Maybe @engdoreis has more experience with UART in the field?

AlexJones0 replied:
Yes, that's definitely possible. I didn't even know this was a possible configuration until I saw that this option was present in opentitanlib. From some brief searching it seems like a rare configuration option that is used by some UARTs when they have e.g. 5/6 data bits.

I think that from our bitbanging side, it should be sufficient for anyone who wants 1.5 bits just to use a 1 bit decoder and a 2 bit encoder, and so I encoded this as just a `u8` for ease. If you or @engdoreis think its important I can switch to using an enum instead and just mark 1.5 bits as unsupported (otherwise I would have to halve the sample period which would introduce a lot more complexity - for similar reasons I haven't yet introduced double data rate in SPI).

engdoreis replied:
I think that enum would be better in this case, even if we never use 1.5 bits, it would constrain the values which is alway good.

pamaury commented on sw/host/opentitanlib/src/test_utils/bitbanging/uart.rs:
```diff
@@ -118,18 +139,209 @@ impl<const TX: u8> UartBitbangEncoder<TX> {
+
+/// Possible states for the decoder state machine to be in.
+#[derive(Debug, PartialEq)]
+enum DecodingState {
```
If would useful to document what the values in each state mean. I had to read the code to understand what they do.

AlexJones0 replied:
Sure, I can do that. In general the values stored with each state are required to be stored to continue parsing in that state (e.g. when parsing stop bits, we need to know how many bits we've parsed so far. When parsing data, we likewise need to know how many bits we've parsed so far). The data in `parsed_data` and `parsed_parity` is stored separate from the values because I need it to persist across decoding states (data, parity, stop) and I thought it would be more unwieldy & verbose to pass it through all the states. I'll try and improve the documentation on these values & fields.

pamaury commented on sw/host/opentitanlib/src/test_utils/bitbanging/uart.rs:
```diff
@@ -118,18 +139,209 @@ impl<const TX: u8> UartBitbangEncoder<TX> {
+    }
+
+    /// A helper to change the parity of the decoder
+    pub fn set_parity(&mut self, parity: Parity) {
```
Why is there specifically a function to change the parity?

AlexJones0 replied:
For supporting the opentitanlib UART trait (for eventual bitbangign work), this is the one attribute that may need to be changed during UART operation with the UART's `set_parity` function. I explicitly added a helper for this with the intention that other similar helpers could be easily added if needed in the future.

AlexJones0 commented on sw/host/opentitanlib/src/test_utils/bitbanging/uart.rs:
```diff
@@ -118,18 +139,209 @@ impl<const TX: u8> UartBitbangEncoder<TX> {
+                    bits: 0,
+                };
+            }
+            DecodingState::Stop { mut data, mut bits } => {
```
As above, I put some parsed data into global state to avoid passing it through the decoding states (and making the interface more verbose), but I could change this to store everything in the state if you think that's nicer.

I was mostly trying to avoid signatures like
`DecodingState::Stop { data: u8, parity: Option<bool>, stop_data: u8, stop_bits: u8 }`
but I'm happy to change to that if you think it's more readable.

AlexJones0 commented:
@pamaury thanks for the review! I've removed the stateful `parsed_data` and `parsed_parity` and retained all the state in the `DecodingState` fields, which I've now properly documented. The code's a bit more verbose now but hopefully should be easier to follow. When you get the chance, can you give it a look and check this matches what you were thinking?

I've also removed the `set_parity` functions for now and will add them to the later follow up PRs that would use them.

pamaury commented on sw/host/opentitanlib/src/test_utils/bitbanging/uart.rs:
```diff
@@ -0,0 +1,658 @@
+                        self.state = DecodingState::Idle;
+                        return Ok(Some(decoded));
+                    }
+                    let mut break_bits = 1 + self.config.data_bits + stop_bits;
```
Isn't that just the frame time?
